crates/swc_plugin_runner/src/cache.rs

```diff
@@ -0,0 +1,153 @@
+use std::{
+    env::current_dir,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{Context, Error};
+use parking_lot::Mutex;
+use swc_common::{
+    collections::AHashMap,
+    sync::{Lazy, OnceCell},
+};
+use wasmer::{Module, Store};
+use wasmer_cache::{Cache as WasmerCache, FileSystemCache, Hash};
+
+/// Version for bytecode cache stored in local filesystem.
+///
+/// This MUST be updated when bump up wasmer.
+///
+/// Bytecode cache generated via wasmer is generally portable,
+/// however it is not gauranteed to be compatible across wasmer's
+/// internal changes.
+/// https://github.com/wasmerio/wasmer/issues/2781
+const MODULE_SERIALIZATION_VERSION: &str = "v1";
+
+/// A shared instance to plugin's module bytecode cache.
+/// TODO: it is unclear how we'll support plugin itself in wasm target of
+/// swc, as well as cache.
+pub static PLUGIN_MODULE_CACHE: Lazy<PluginModuleCache> = Lazy::new(Default::default);
+
+#[derive(Default)]
+pub struct CacheInner {
+    fs_cache: Option<FileSystemCache>,
+    memory_cache: InMemoryCache,
+}
+
+/// Lightweight in-memory cache to hold plugin module instances.
+/// Current it doesn't have any invalidation or expiration logics like lru,
+/// having a lot of plugins may create some memory pressure.
+#[derive(Default)]
+pub struct InMemoryCache {
+    modules: AHashMap<PathBuf, Module>,
+}
+
+#[derive(Default)]
+pub struct PluginModuleCache {
+    inner: OnceCell<Mutex<CacheInner>>,
+    /// To prevent concurrent access to `WasmerInstance::new`
+    instantiation_lock: Mutex<()>,
+}
+
+fn create_filesystem_cache(filesystem_cache_root: &Option<String>) -> Option<FileSystemCache> {
+    let mut root_path = if let Some(root) = filesystem_cache_root {
+        Some(PathBuf::from(root))
+    } else if let Ok(cwd) = current_dir() {
```

Comment:
This will use wrong directory if user manually invokes `npx swc` from non-root directory.
Maybe we need to use directory of `package.json` or `.swcrc`

Reply:
`filesystem_cache_root` is a configuration option from `.swcrc` already, this is for the default path. 

We do try to create default from cwd and gracefully falls back to not create cache path if cwd is not accessible. npx is usually treated as one-off, on non-CI execution path which would be ok without fs cache support.